UmpleParser/src/ParseUtilities_Code.ump

```diff
@@ -0,0 +1,606 @@
+    }
+  }
+
+  public int getIntValue(String tokenName)
+  {
+    return getIntValue(tokenName,0);
+  }
+
+  public int getIntValue(String tokenName, int defaultIfInvalid)
+  {
+    String value = getValue(tokenName);
+    if (value == null)
+    {
+      return defaultIfInvalid;
+    }
+    else
+    {
+      try
+      {
+        return Integer.parseInt(value);
+      }
+      catch (NumberFormatException e)
+      {
+        return defaultIfInvalid;
+      }
+    }
+  }
+
+  public Token getSubToken(String tokenName, int offset)
+  {
+    if (is(tokenName))
+    {
+      return this;
+    }
+    else
+    {
+      int currentOffset = 0;
+      for (Token t : subTokens)
+      {
+        if (t.is(tokenName))
+        {
+          if (currentOffset == offset)
+          {
+            return t;  
+          }
+          currentOffset += 1;  
+        }
+      }
+    }
+    return null;
+  }
+
+  public Token getSubToken(String tokenName)
+  {
+    return getSubToken(tokenName,0);
+  }
+
+  public String getValue(String tokenName)
+  {
+    Token t = getSubToken(tokenName);
+    if (t != null)
+    {
+      return t.getValue();
+    }
+    else
+    {
+      // This should be "", as getValue() returns "" by default
+      return null;
+    }
+  }
+
+  public boolean is(String aName)
+  {
+    return getName().equals(aName) && !isStatic();
+  }
+
+  public boolean isStatic(String aName)
+  {
+    return getName().equals(aName) && isStatic();
+  }
+
+  public boolean isStatic()
+  {
+    return "STATIC".equals(getValue());
+  }
+
+  public boolean isValue(String compareTo)
+  {
+    return getValue().equals(compareTo);
+  }
+
+  public void remove(int index)
+  {
+    if (index < 0 || index >= subTokens.size())
+    {
+      return;
+    }
+    subTokens.remove(index);
+  }
+
+  public String toString()
+  {
+    StringBuffer answer = new StringBuffer();
+    toString(answer,null);
+    return answer.toString();
+  }
+
+  // Code refactored from Parser class to Token class
+  public StringBuffer toString(StringBuffer stringSoFar, Token parserRoot)
+  {
+    if ("START_TOKEN".equals(getValue()))
+    {
+      stringSoFar.append("[" + getName() + "]");
+    }
+    else if ("STATIC".equals(getValue()))
+    {
+      //ignore
+    }
+    else if (parserRoot == null || !this.equals(parserRoot))
+    {
+      stringSoFar.append("[" + getName() + ":" + getValue() + "]");
+    }
+
+    for(Token subToken : getSubTokens())
+    {
+      subToken.toString(stringSoFar,null);
+    }
+    return stringSoFar;
+  }
+  
+  // Return a copy of the current token with all its SubTokens
+  // Removes all SubTokens from the current Token and sets new name and value
+  public Token detach(String newName, String newValue){
+    Token nToken = this.shallowCopy();
+    setName(newName);
+    setValue(newValue);
+    ArrayList<Token> subTokenList = new ArrayList<Token>(subTokens);
```

Comment:
I don't really get what this line is doing. It's cloning something, but the `subTokenList ` variable is never modified. Thus, it's a needless copy. 

The next for loop should just iterate over `subTokens`.